R/nm-join-bayes.R

```diff
@@ -0,0 +1,583 @@
+  )
+}
+
+#' Prepare for joining data to table results from submodels
+#'
+#' For each submodule, join the data and specified tables with `bbr::nm_join()`,
+#' and summarize the table values across chains with the specified point
+#' function.
+#'
+#' This function does *not* do the final join; instead it returns a list with
+#' two fields, `data` and `tab` (summarized table values), so that the calling
+#' code doesn't need to do extra work to figure out what the data columns are.
+#'
+#' @noRd
+prepare_join <- function(mod, join_col, files, superset, point_fn, ...) {
+  join_args <- list(.join_col = join_col, .superset = superset, ...)
+  if (!is.null(files)) {
+    join_args <- c(join_args, .files = files)
+  }
+
+  join <- function(sdir) {
+    do.call(bbr::nm_join, c(list(read_model(sdir)), join_args))
+  }
+
+  sdirs <- get_chain_dirs(mod)
+  nchains <- length(sdirs)
+  if (nchains < 2) {
+    stop("Valid nmbayes models need more than one chain.")
+  }
+
+  # The result of these bbr::nm_join() calls doesn't depend on the RNG, but 1)
+  # processx::process$new() moves the seed (tested with processx v3.8.2) and 2)
+  # a variable number of calls to that are triggered by bbr::nm_join() depending
+  # on whether bbi_exec() has been called yet in the session. Restore seed on
+  # exit so that the downstream simulation results don't depend on these
+  # details.
+  withr::local_preserve_seed()
+
+  dfs <- vector(mode = "list", length = nchains)
+  # Do first call separately to honor caller's bbr.verbose setting.
+  dfs[[1]] <- join(sdirs[1])
+  verbose_msg("\nRepeating join for other chains and combining results.")
+  withr::with_options(list(bbr.verbose = FALSE), {
+    for (i in 2:nchains) {
+      dfs[[i]] <- join(sdirs[i])
+    }
+  })
+
+  # nm_join_origin records the source of different columns.
+  origin <- attr(dfs[[1]], "nm_join_origin")
+  if (is.null(origin)) {
+    # For older bbr versions, fall back to peeking at the header and making the
+    # same transformations as nm_join/nm_data/nm_file.
+    #
+    # TODO: Remove when minimum bbr version is above 1.7.0.
+    smod1_data <- bbr::get_data_path(read_model(get_chain_dirs(mod)[1]))
+    data_cols <- toupper(fread_peek_at_columns(smod1_data, skip = join_col))
+
+    dv <- data_cols == "DV"
+    if (any(dv)) {
+      data_cols[dv] <- "DV.DATA"
+    }
+  } else {
+    data_cols <- origin$data
+  }
+
+  data <- dfs[[1]][data_cols]
+
+  chain_col <- ".bbr_bayes_nm_join_chain"
+  tab_sum <- dplyr::bind_rows(dfs, .id = chain_col) %>%
+    dplyr::group_by(.data[[join_col]]) %>%
+    dplyr::summarize(
+      dplyr::across(-any_of(c(data_cols, chain_col)), .fns = point_fn)
+    )
+
+  if (nrow(data) != nrow(tab_sum)) {
+    stop("bug: number of data rows not same as number of table summary rows")
+  }
+
+  return(list(data = data, tab = tab_sum))
+}
+
+call_presim <- function(fn, x, join_col) {
+  nrow_orig <- nrow(x)
+  x <- fn(x)
+
+  if (!inherits(x, "data.frame")) {
```

Comment:
These checks all look good to me. We're kind of getting into deep water here with letting the user pass in a function, but if the results passes these checks... I guess I think we're doing our due diligence.